syntax = "proto3";

package astraios.compile;

import "protos/tierkreis/graph.proto";

message CompileRequest {
  // Map from CellIds to code
  map<string, string> cell_contents = 1;
  // Worker to submit compiled code to
  string worker_id = 2;
  // Map of known variables and their types
  map<string, tierkreis.v1alpha1.graph.Type> scope = 3;
}

message Variable {
    string name = 1;
    tierkreis.v1alpha1.graph.Type type = 2;
}

message CompiledFunction {
    string func_id = 1;
    string cell_id = 2;
    repeated string inputs = 3;
    repeated string outputs = 4;
    map<string, Variable> variables = 5;
}

message CompileResult {
    // Map from CellIds to Compiled functions
    map<string, CompiledFunction> func_ids = 1;
}

message CompileStatus {
    string status = 1;
}

message CompileError {
    string error = 1;
}

message CompileResponse {
    oneof Response {
        CompileResult result = 1;
        CompileStatus status = 2;
    }
}

service Compilation {
    rpc Compile(CompileRequest) returns (stream CompileResponse);
}

